*** configure	2020-02-19 15:55:05.049975110 -0800
--- configure_new	2020-02-19 15:53:58.853492011 -0800
***************
*** 182,188 ****
  if test "$gcc" -eq 1 && ($cc -c $test.c) >> configure.log 2>&1; then
    echo ... using gcc >> configure.log
    CC="$cc"
!   CFLAGS="${CFLAGS--O3}"
    SFLAGS="${CFLAGS--O3} -fPIC"
    if test "$ARCHS"; then
      CFLAGS="${CFLAGS} ${ARCHS}"
--- 182,188 ----
  if test "$gcc" -eq 1 && ($cc -c $test.c) >> configure.log 2>&1; then
    echo ... using gcc >> configure.log
    CC="$cc"
!   CFLAGS="${CFLAGS--O3} -fPIC -O3"
    SFLAGS="${CFLAGS--O3} -fPIC"
    if test "$ARCHS"; then
      CFLAGS="${CFLAGS} ${ARCHS}"
